edition = "2023";

service Coordinator {
  rpc RegisterWorkerProcess(WorkerProcessID) returns (Empty) {}
  rpc UnregisterWorkerProcess(WorkerProcessID) returns (Empty) {}

  rpc GetNextTask(WorkerProcessID) returns (TaskAssignment) {}
  rpc SetTaskResult(TaskResult) returns (Empty) {}

  rpc QueueTryPut(QueuePutRequest) returns (QueuePutResponse) {}
  rpc QueueTryGet(QueueID) returns (QueueGetResponse) {}
  rpc QueueShutdown(QueueID) returns (Empty) {}
}

message Empty {}

message WorkerProcessID {
  string group = 1;
  string name = 2;
  int64 slurm_job_id = 3;
  string hostname = 4;
  int64 pid = 5;
}

message TaskDefn {
  string task_id = 1;
  FunctionCall function_call = 2;
}

message FunctionCall {
  bytes function = 1;
  bytes args = 2;
  bytes kwargs = 3;
}

message TaskAssignment {
  oneof assignment {
    bool exit_flag = 1;
    TaskDefn task = 2;
  }
}

message TaskResult {
  string task_id = 1;
  oneof result {
    bytes retval = 2;
    Error error = 3;
  }
  WorkerProcessID process_id = 4;
  double loads_input_duration = 5;
  double run_duration = 6;
  double dumps_output_duration = 7;
}

message Error {
  string message = 1;
  string error_id = 2;
}

enum QueueErrorCode {
  FULL = 0;
  EMPTY = 1;
  SHUTDOWN = 2;
}

message QueueID { string queue = 1; }

message QueuePutRequest {
  string queue = 1;
  bytes item = 2;
}

message QueuePutResponse { QueueErrorCode error = 1; }

message QueueGetResponse {
  oneof result {
    bytes item = 1;
    QueueErrorCode error = 2;
  }
}
